### ISSUE TITLE
PDF rasterized when instructed not to

### ISSUE THREAD

**Issue by klabarge** [qzind/tray#278]
A client is experiencing an issue with PDF's that grow very large in the printer queue (700MB) that result in very slow print speeds. Reproducing the issue exposed that that vector PDF was being rasterized despite `rasterize: false`

### Environment

**OS:** Windows 10
**QZ Tray:** 2.0.5
**Java:** 8u161 64-bit

* [Here is the offending PDF](https://github.com/qzind/tray/files/1749977/test.pdf) that can be used to reproduce the exact issue

**UPDATE:**  The basic PDF ([the underlying form/template attached here](https://github.com/qzind/tray/files/1750080/template.pdf)) was made with Adobe Illustrator, the filled data is another PDF; merged together with [FPDI](https://www.setasign.com/products/fpdi/about/).

## Expected Results
![screen shot 2018-02-22 at 9 52 57 pm](https://user-images.githubusercontent.com/6345473/36575790-d730b36e-181a-11e8-8eea-855b05383b2e.png)

### PDFCreator

1. When `rasterize` is `false`, the print is a rasterized image.
    ![image](https://user-images.githubusercontent.com/12505463/36571015-c1ea7d5e-1802-11e8-8bfb-84983dd0479a.png)

1. Printing the offending PDF using the virtual printer PDFCreator results in a 700MB print job.
   * Once spooled, the file is about 6.5MB on disk

   ![image](https://user-images.githubusercontent.com/12505463/36570826-e6b6a99c-1801-11e8-99ad-ee8b9ffb0fb5.png)

### Microsoft Print to PDF

1. When `rasterize` is `false`, the print is a rasterized image.

   ![image](https://user-images.githubusercontent.com/12505463/36573252-e890d4e8-180d-11e8-97bf-96fbacb28eb1.png)

1. I was unable to reproduce the large print job using the Microsoft Print to PDF printer
   ![image](https://user-images.githubusercontent.com/12505463/36573218-beaa0f96-180d-11e8-840b-8d3d9fca3aca.png)

### Stacktrace

```
[DEBUG] 2018-02-22 20:14:31,692 @ qz.ws.PrintSocketClient:?
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"Microsoft Print to PDF"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":null,"interpolation":"","jobName":"","margins":"0","orientation":"","paperThickness":"","printerTray":"","rasterize":false,"rotation":"0","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"pdf","data":"file:///C:/Users/Kyle/Desktop/demo/assets/test.pdf"}]},"timestamp":1519348471690,"uid":"uq3k7v"}
[WARN] 2018-02-22 20:14:31,697 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2018-02-22 20:14:32,695 @ qz.common.TrayManager:?
	Allowed localhost to print to Microsoft Print to PDF
[TRACE] 2018-02-22 20:14:32,922 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/1 already in use
[DEBUG] 2018-02-22 20:14:32,922 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintPDF to print
[DEBUG] 2018-02-22 20:14:32,937 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching Microsoft Print to PDF
[DEBUG] 2018-02-22 20:14:32,937 @ qz.printer.PrintServiceMatcher:?
	Found 4 printers
[DEBUG] 2018-02-22 20:14:32,937 @ qz.printer.PrintServiceMatcher:?
	Found match: Microsoft Print to PDF
[WARN] 2018-02-22 20:14:32,937 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2018-02-22 20:14:32,937 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2018-02-22 20:14:32,937 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for orientation, using default
[WARN] 2018-02-22 20:14:32,937 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[DEBUG] 2018-02-22 20:14:32,952 @ qz.utils.PrintingUtilities:?
	Found Resolution: 60000x60000 dphi
[DEBUG] 2018-02-22 20:14:33,467 @ qz.printer.action.PrintPDF:?
	Parsed 4 files for printing
[TRACE] 2018-02-22 20:14:33,576 @ qz.printer.action.PrintPixel:?
	DPI: 0.0	CNV: 72.0
[TRACE] 2018-02-22 20:14:33,576 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2018-02-22 20:14:33,576 @ qz.printer.action.PrintPixel:?
	[(0.0,0.0)->(215.9,279.4)mm, color]
[TRACE] 2018-02-22 20:14:33,684 @ qz.printer.action.PrintPixel:?
	DPI: 0.0	CNV: 72.0
[TRACE] 2018-02-22 20:14:33,793 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2018-02-22 20:14:33,809 @ qz.printer.action.PrintPixel:?
	[(0.0,0.0)->(215.9,279.4)mm, color]
[TRACE] 2018-02-22 20:14:33,839 @ qz.printer.action.PrintPixel:?
	DPI: 0.0	CNV: 72.0
[TRACE] 2018-02-22 20:14:33,839 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2018-02-22 20:14:33,839 @ qz.printer.action.PrintPixel:?
	[(0.0,0.0)->(215.9,279.4)mm, color]
[TRACE] 2018-02-22 20:14:33,869 @ qz.printer.action.PrintPixel:?
	DPI: 0.0	CNV: 72.0
[TRACE] 2018-02-22 20:14:33,869 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2018-02-22 20:14:33,869 @ qz.printer.action.PrintPixel:?
	[(0.0,0.0)->(215.9,279.4)mm, color]
[TRACE] 2018-02-22 20:14:33,900 @ qz.printer.action.PrintPixel:?
	DPI: 0.0	CNV: 72.0
[TRACE] 2018-02-22 20:14:33,900 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2018-02-22 20:14:33,900 @ qz.printer.action.PrintPixel:?
	[(0.0,0.0)->(215.9,279.4)mm, color]
[INFO] 2018-02-22 20:14:33,900 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[TRACE] 2018-02-22 20:14:42,955 @ qz.printer.BookBundle:?
	Requested page 0 for printing
[TRACE] 2018-02-22 20:14:42,955 @ qz.printer.BookBundle:?
	Requested page 0 for printing
[DEBUG] 2018-02-22 20:14:42,955 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[WARN] 2018-02-22 20:14:43,198 @ org.apache.pdfbox.pdmodel.font.PDSimpleFont:325
	No Unicode mapping for gid00422 (31) in font MFWBCI+SourceSerifVariable-Roman_600_
[TRACE] 2018-02-22 20:14:43,961 @ qz.printer.BookBundle:?
	Requested page 0 for printing
[DEBUG] 2018-02-22 20:14:43,977 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:45,151 @ qz.printer.BookBundle:?
	Requested page 0 for printing
[DEBUG] 2018-02-22 20:14:45,151 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:47,550 @ qz.printer.BookBundle:?
	Requested page 1 for printing
[TRACE] 2018-02-22 20:14:47,566 @ qz.printer.BookBundle:?
	Requested page 1 for printing
[DEBUG] 2018-02-22 20:14:47,566 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[WARN] 2018-02-22 20:14:47,863 @ org.apache.pdfbox.pdmodel.font.PDSimpleFont:325
	No Unicode mapping for gid00423 (30) in font MFWBCI+SourceSerifVariable-Roman_600_
[TRACE] 2018-02-22 20:14:48,394 @ qz.printer.BookBundle:?
	Requested page 1 for printing
[DEBUG] 2018-02-22 20:14:48,410 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:49,566 @ qz.printer.BookBundle:?
	Requested page 1 for printing
[DEBUG] 2018-02-22 20:14:49,582 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:51,832 @ qz.printer.BookBundle:?
	Requested page 2 for printing
[TRACE] 2018-02-22 20:14:51,847 @ qz.printer.BookBundle:?
	Requested page 2 for printing
[DEBUG] 2018-02-22 20:14:51,847 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[WARN] 2018-02-22 20:14:52,160 @ org.apache.pdfbox.pdmodel.font.PDSimpleFont:325
	No Unicode mapping for gid00424 (29) in font MFWBCI+SourceSerifVariable-Roman_600_
[TRACE] 2018-02-22 20:14:52,644 @ qz.printer.BookBundle:?
	Requested page 2 for printing
[DEBUG] 2018-02-22 20:14:52,644 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:53,425 @ qz.printer.BookBundle:?
	Requested page 2 for printing
[DEBUG] 2018-02-22 20:14:53,425 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:55,519 @ qz.printer.BookBundle:?
	Requested page 3 for printing
[TRACE] 2018-02-22 20:14:55,535 @ qz.printer.BookBundle:?
	Requested page 3 for printing
[DEBUG] 2018-02-22 20:14:55,535 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[WARN] 2018-02-22 20:14:55,785 @ org.apache.pdfbox.pdmodel.font.PDSimpleFont:325
	No Unicode mapping for gid00425 (28) in font MFWBCI+SourceSerifVariable-Roman_600_
[TRACE] 2018-02-22 20:14:56,176 @ qz.printer.BookBundle:?
	Requested page 3 for printing
[DEBUG] 2018-02-22 20:14:56,178 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[TRACE] 2018-02-22 20:14:56,909 @ qz.printer.BookBundle:?
	Requested page 3 for printing
[DEBUG] 2018-02-22 20:14:56,924 @ qz.printer.PDFWrapper:?
	Paper area: 0,0:612,792
[INFO] 2018-02-22 20:14:58,943 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2018-02-22 20:14:58,959 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**Comment by klabarge** [comment]
This may be related to #279

The offending PDF in #279 is the PDF template for the PDF in this bug report.

**Comment by tresf** [comment]
Confirmed on MacOS as well.  It appears CUPS-PDF has a very low default resolution so the issue is per

```bash
java -jar /Users/owner/Downloads/pdfbox-app-2.0.8.jar PrintPDF -silentPrint -defaultPrinter CUPS-PDF -dpi 0 /Users/tresf/Downloads/test.pdf
```

![screen shot 2018-02-22 at 9 47 28 pm](https://user-images.githubusercontent.com/6345473/36575645-09d495fc-181a-11e8-9312-4c2e79231b24.png)

**Comment by tresf** [comment]
Filed upstream bug report [`PDFBOX-4123`](https://issues.apache.org/jira/browse/PDFBOX-4123).

**Comment by tresf** [comment]
Explanation from [`PDFBOX-4123(comment)`](https://issues.apache.org/jira/browse/PDFBOX-4123?focusedCommentId=16375570&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16375570):

> *Your PDF has transparency, this results in Java printing it as a bitmap... Here's a [competitor] ... that mentions this:*

> > *[...] "PostScript is the grandaddy of PDF, and if you're reading about printing you've already heard of it. It can represent text, vector graphics and images, so if your PDF contains a vector shape it's going to be reproduced almost exactly in PostScript. **Other graphics operations (like transparency) can't be done in PostScript, and the Java print layer handles this by rasterizing the page to a bitmap and including that bitmap in the spool file.**" - [© 2012 Jonathan Cookson, bfo.com](http://bfo.com/blog/2012/02/15/using_java_to_print_pdf_documents/)*

**Comment by tresf** [comment]
So anyone wanting to hack transparency out, there's a way to do it on a binary level... here's the reply from Tilman:

> *"I just did, by replacing `/Group with `/Xroup which makes it inactive, i.e. just an ordinary XObject form and no longer a transparency group
> [...]
> I did it with NOTEPAD++.
> There is no direct simple technique with PDFBox, but you could go through each dictionary (page, but also resources like `XObject` forms and patterns) and then remove it on the `COS` level.
> For PDF files like the one attached it makes no difference (or I can't see it), but for "nice" PDFs (e.g. illustrations) the result would be very different.
> The better solution would be not to use transparency groups when creating the PDF. Your PDF is just an average looking form."  - Tilman Hausherr 2018-02-25 @PDFBOX-4123*

Closing, marking as `wontfix`.

**Comment by tresf** [comment]
> So anyone wanting to hack transparency out, there's a way to do it on a binary level... here's the reply from Tilman:
> 
> > _"I just did, by replacing `/Group` with `/Xroup` which makes it inactive, i.e. just an ordinary XObject form and no longer a transparency group
> > [...]
> > I did it with NOTEPAD++.
> > There is no direct simple technique with PDFBox, but you could go through each dictionary (page, but also resources like `XObject` forms and patterns) and then remove it on the `COS` level.
> > For PDF files like the one attached it makes no difference (or I can't see it), but for "nice" PDFs (e.g. illustrations) the result would be very different.
> > The better solution would be not to use transparency groups when creating the PDF. Your PDF is just an average looking form."  - Tilman Hausherr 2018-02-25 @PDFBOX-4123_
> 
> Closing, marking as `wontfix`.

Turns out we have an even better way... 

https://github.com/qzind/tray/pull/959#issuecomment-1112823274

**Comment by tresf** [comment]
Documentation published https://qz.io/wiki/pixel#pdf-transparency.   This feature will be available starting with QZ Tray 2.2.2.
